WordPress/Classes/Utility/WPAuthTokenIssueSolver.m

```diff
@@ -0,0 +1,133 @@
+        loginViewController.cancellable = ![self noSelfHostedBlogs];
+        loginViewController.dismissBlock = ^(BOOL cancelled) {
+            if (cancelled) {
+                [self showCancelReAuthenticationAlertAndOnOK:^{
+                    NSManagedObjectContext *mainContext = [[ContextManager sharedInstance] mainContext];
+                    AccountService *accountService = [[AccountService alloc] initWithManagedObjectContext:mainContext];
+                    
+                    [accountService removeDefaultWordPressComAccount];
+                    onComplete();
+                }];
+            } else {
+                onComplete();
+            }
+        };
+        
+        [UIApplication sharedApplication].keyWindow.rootViewController = loginViewController;
+        
+        [self showExplanationAlertForReAuthenticationDueToMissingAuthToken];
+        isFixingAuthTokenIssue = YES;
+    } else {
+        onComplete();
+    }
+    
+    return isFixingAuthTokenIssue;
+}
+
+#pragma mark - Misc
+
+/**
+ *  @brief      Call this method to know if there are hosted blogs.
+ *
+ *  @returns    YES if there are hosted blogs, NO otherwise.
+ */
+- (BOOL)noSelfHostedBlogs
+{
+    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
+    BlogService *blogService = [[BlogService alloc] initWithManagedObjectContext:context];
+    
+    NSInteger blogCount = [blogService blogCountSelfHosted];
+    return blogCount == 0;
+}
+
+/**
+ *  @brief      Call this method to know if the local installation of WPiOS has the authToken issue
+ *              this class was designed to solve.
+ *
+ *  @returns    YES if the local WPiOS installation needs to be fixed by this class.
+ */
+- (BOOL)hasAuthTokenIssues
+{
+    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
+    AccountService *accountService = [[AccountService alloc] initWithManagedObjectContext:context];
+    WPAccount *account = [accountService defaultWordPressComAccount];
+    
+    BOOL hasAuthTokenIssues = account && ![account authToken];
+    
+    return hasAuthTokenIssues;
+}
+
+#pragma mark - Alerts
+
+/**
+ *  @brief      Shows the alert when the re-authentication is cancelled by the user.
+ *
+ *  @param      okBlock     The block that will be executed if the user confirms the operation.
+ */
+- (void)showCancelReAuthenticationAlertAndOnOK:(WPAuthTokenissueSolverCompletionBlock)okBlock
+{
+    NSParameterAssert(okBlock);
+    
+    NSString *alertTitle = NSLocalizedString(@"Careful!",
+                                             @"Title for the warning shown to the user when he refuses to re-login when the authToken is missing.");
+    NSString *alertMessage = NSLocalizedString(@"If you proceed, all your WP.com account data will be removed from this device.  This means any unsaved data such as locally stored posts will be deleted.  This will not affect data that has already been uploaded to your WP.com account.",
```

Comment:
We should spell out WordPress.com. Also maybe tweak the message to be a bit more friendly and less technical. Maybe something like:

> Proceeding will remove all WordPress.com data from this device, and delete any locally saved drafts.  You will not loose anything already saved to your WordPress.com blog(s).